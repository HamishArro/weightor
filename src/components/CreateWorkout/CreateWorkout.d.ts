import {Exercises, Cardio} from '../../utils/Workout/Workout';

export type CreateProps = {
  title: string;
  onPress: () => void;
};

export type DisplayProps = {
  exercises: Exercises;
  addButtonText: string;
  addHandler: () => void;
};

export type AddProps = {
  addCardio: (exercise: Cardio) => void;
};
