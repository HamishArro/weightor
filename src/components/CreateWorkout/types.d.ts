import {Exercises, Cardio} from '../../hooks/useWorkout/types';

export type CreateProps = {
  title: string;
  onPress: () => void;
};

export type DisplayProps = {
  exercises: Exercises;
  addButtonText: string;
  handleAdd: () => void;
};

export type AddProps = {
  handleBack: () => void;
  handleCardio: () => void;
  handleWeight: () => void;
};

export type CardioProps = {
  handleBack: () => void;
  handleAdd: (exercise: Cardio) => void;
};
