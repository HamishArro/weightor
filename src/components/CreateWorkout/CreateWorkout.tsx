import React, {useState} from 'react';
import {View} from 'react-native';
import useWorkout, {
  Cardio as CardioType,
} from '../../hooks/useWorkout/useWorkout';
import Create from './Create/Create';
import Display from './Display/Display';
import Cardio from './Cardio/Cardio';
import Add from './Add/Add';
import {styles} from '../../utils/styleSheet';

enum states {
  'ADD',
  'CREATE',
  'CARDIO',
  'DISPLAY',
}

function CreateWorkout(): JSX.Element {
  const [flowState, setFlowState] = useState<states>(states.CREATE);

  const {exercises, addCardioExercise} = useWorkout();

  const handleAdd = (exercise: CardioType) => {
    addCardioExercise(exercise);
    setFlowState(states.DISPLAY);
  };

  const handleBack = exercises.length
    ? () => setFlowState(states.DISPLAY)
    : () => setFlowState(states.CREATE);

  const stateSwitch = () => {
    switch (flowState) {
      case states.ADD:
        return (
          <Add
            handleBack={handleBack}
            handleCardio={() => setFlowState(states.CARDIO)}
            handleWeight={() => setFlowState(states.CARDIO)}
          />
        );
      case states.CREATE:
        return (
          <Create title="Create" onPress={() => setFlowState(states.ADD)} />
        );
      case states.CARDIO:
        return (
          <Cardio
            handleAdd={handleAdd}
            handleBack={() => setFlowState(states.ADD)}
          />
        );
      case states.DISPLAY:
        return (
          <Display
            exercises={exercises}
            addButtonText="Add another exercise"
            handleAdd={() => setFlowState(states.ADD)}
          />
        );
    }
  };

  return (
    <View style={styles.container} testID="create-workflow-container">
      {stateSwitch()}
    </View>
  );
}

export default CreateWorkout;
