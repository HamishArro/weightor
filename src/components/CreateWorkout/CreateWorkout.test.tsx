import * as React from 'react';
import {describe, it, beforeEach} from '@jest/globals';
import {fireEvent, render, screen, act} from '@testing-library/react-native';
import CreateWorkout from './CreateWorkout';

describe('create workout tests', () => {
  beforeEach(() => {
    render(<CreateWorkout />);
  });

  it(`will take me to create when I haven't added an exercise`, () => {
    act(() => fireEvent.press(screen.getByTestId('create-button')));

    screen.getByTestId('add-container');

    act(() => fireEvent.press(screen.getByTestId('back-button')));

    screen.getByTestId('create-button');
  });

  it(`will take me to add when I click back on cardio`, () => {
    act(() => fireEvent.press(screen.getByTestId('create-button')));

    screen.getByTestId('add-container');

    act(() => fireEvent.press(screen.getByTestId('cardio-button')));

    screen.getByTestId('cardio-container');

    act(() => fireEvent.press(screen.getByTestId('back-button')));

    screen.getByTestId('add-container');
  });

  it(`can create a weight exercise successfully`, () => {
    act(() => fireEvent.press(screen.getByTestId('create-button')));

    screen.getByTestId('add-container');

    act(() => fireEvent.press(screen.getByTestId('weight-button')));
  });

  it('can create a cardio workout successfully', async () => {
    act(() => fireEvent.press(screen.getByTestId('create-button')));

    act(() => fireEvent.press(screen.getByTestId('cardio-button')));

    act(() => {
      fireEvent.changeText(screen.getByTestId('name-input'), 'light jog');
      fireEvent.changeText(screen.getByTestId('muscles-used-input'), 'legs');
      fireEvent.changeText(screen.getByTestId('workout-effort-input'), '80');
      fireEvent.changeText(screen.getByTestId('duration-input'), '900');
    });

    act(() => fireEvent.press(screen.getByTestId('add-button')));

    screen.getByTestId('display-container');

    act(() => fireEvent.press(screen.getByTestId('add-button')));

    screen.getByTestId('add-container');

    act(() => fireEvent.press(screen.getByTestId('back-button')));

    screen.getByTestId('display-container');
  });
});
