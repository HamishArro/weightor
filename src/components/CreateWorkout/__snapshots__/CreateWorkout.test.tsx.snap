// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create workout tests matches snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "margin": 32,
    }
  }
  testID="create-workflow-container"
>
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="create-button"
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        create
      </Text>
    </View>
  </View>
</View>
`;
